### PR TITLE
treewide: expunge functions from platforms

### PR DESCRIPTION
**Pull request by ghost**
Cc: @roberth @Artturin @sternenseemann 

###### Description of changes

This commit moves `canExecute`, `emulator`, and `emulatorAvailable` out of `{host,build,target}Platform` and lifts them into `lib.systems`.  Stubs have been left behind to assist with migration of out-of-tree code.

Nix's function equality is unpredictable and error-prone; we should not rely on it.  Since we need to be able to test `{host,build,target}Platform` attrsets for equality, the most sustainable long-term solution is to forbid functions from these attrsets, much like we do for the `meta` attrset.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: lib/systems/default.nix
Comment (emilazy) [inline lib/systems/default.nix:116]: it's 2023 :)

**Review by sternenseemann — APPROVED** [review]
Excellent!

* With 5d8e668d16f7ce7c32399007d3a49a27322ac3ed, the issue described in #237427 should disappear and we'd be able to revert #237512 for good (and live worriless when the functions are finally removed after a deprecation period of one or two releases, I guess).
* We should add polyfills to 23.11, so backporting cross changes becomes easier. These would have the form: `canExecute = machinePlatform: machinePlatform.canExecute` etc. This wouldn't break anything, but would ideally make backported changes just work (as long as they apply)

File: pkgs/top-level/all-packages.nix
Comment (Artturin) [reply pkgs/top-level/all-packages.nix:5537]: I added mesonEmulatorHook and it has to be targetPlatform, otherwise it'll always throw. 

AFAIK it's been working well with no issues for about a year

File: lib/systems/default.nix
Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix]: ed5a899d3e9f

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (amjoseph-nixpkgs) [reply nixos/doc/manual/release-notes/rl-2311.section.md]: b972c68905c1

File: lib/systems/default.nix
Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:24]: https://github.com/NixOS/nixpkgs/pull/238480

**Comment by amjoseph-nixpkgs** [comment]
> * With [5d8e668](https://github.com/NixOS/nixpkgs/commit/5d8e668d16f7ce7c32399007d3a49a27322ac3ed), the issue described

Wow nice catch.  That is a great example of how unintuitive Nix's function equality is.  It depends on equality of the entire scope, including variables that aren't mentioned in the expressions being compared.  Which means a lot of basic optimizations will change how it behaves.  Bleh.

**Comment by amjoseph-nixpkgs** [comment]
Squashed down to two commits.

**Comment by emilazy** [comment]
Maybe it'd be better to hoist the placeholder functions outside of the top-level function in their own `let` to ensure they're in a consistent empty scope?

(It'd be great if we could set some flag to make all functions that appear in nixpkgs error on comparison without breaking user code...)

File: pkgs/top-level/all-packages.nix
Comment (amjoseph-nixpkgs) [reply pkgs/top-level/all-packages.nix:5537]: > Unrelated to this PR, but why does `targetPlatform` appear here, but nowhere else relating to `mesonEmulatorHook`? Seems to me this should also be `hostPlatform`.

It seems weird to me, but nixpkgs thinks of things like `pkg-config` and `meson` as having a `targetPlatform`.  So `mesonEmulatorHook`'s `targetPlatform` is the `hostPlatform` of the package that lists `mesonEmulatorHook` as a `nativeBuildInput`.

Pretend for a moment that `build.ninja` is a compiler (which itself is "compiled" by `meson`) and it might make sense.

IMHO if a package doesn't emit binaries, it should be built with `stdenv.targetPlatform==null` in order to eliminate an unnecessary degree of freedom.  But getting to that point is a big, long project.

**Comment by sternenseemann** [comment]
> Maybe it'd be better to hoist the placeholder functions outside of the top-level function in their own `let` to ensure they're in a consistent empty scope?

I guess that'd make the import cache ensure that it'd work even if `lib/systems/default.nix` is imported multiple times, but that doesn't happen in normal circumstances anyways afaik.

**Comment by infinisil** [comment]
Making Nixpkgs rely on the broken function equality sounds like a terrible idea.. what's the intended migration plan away from that?

**Comment by emilazy** [comment]
Nixpkgs already relies on broken function equality. This PR makes it rely on it less; the replaced functions are there only as backwards-compatibility warnings and can presumably be dropped in the future.

File: pkgs/top-level/all-packages.nix
Comment (sternenseemann) [reply pkgs/top-level/all-packages.nix]: Figured out what was irking me, let's discuss at https://github.com/NixOS/nixpkgs/pull/238654.

**Review by roberth — CHANGES_REQUESTED** [review]
I humbly request to stay compatible with 23.05 until its EOL in December.
While this is a serious problem and I agree with the solution, it creates a breaking change for third-party packagers, who will have to add conditionals to deal with the release vs unstable.

We should not break the ecosystem unless it's necessary, and in this case, like most cases, it isn't.

With fixes like https://github.com/NixOS/nixpkgs/pull/237427 and https://github.com/NixOS/nixpkgs/pull/237512, we can ensure that the "bad" equality actually works in the cases that we care about, and we're lucky enough that our usage of these values _is_ very restricted.

Now as for the plan specifically, my main requirement is that someone who maintains a repo with a package that cares about cross compilation will have the option to switch _completely_ as soon as they learn that they need to make a switch. This means that all supported releases (and unstable) must support the new way of doing things before they make the change.

Essentially what that means for this change is to add the new functions to `lib`, but keep the functions on the platform values in place until the 23.05 is EOL.

The same restriction should apply to "mere" warnings as well. In practice a warning is at least as bad for a third-party maintainer, because users will complain about it, and if they can't act on the warning yet, they're stuck with an unsolvable issue.

**Comment by sternenseemann** [comment]
> it creates a breaking change for third-party packagers, who will have to add conditionals to deal with the release vs unstable.

We can offer them a way to do so without breaking compatibility to 23.05 nor adding conditionals by adding the following functions to `lib.systems` in 23.05. This would be a simple solution without pushing taking action to 24.05.

```nix
{
  canExecute = builtins.getAttr "canExecute";
  emulatorAvailable = builtins.getAttr "emulatorAvailable";
  emulator = builtins.getAttr "emulator";
}
```

With this your requirement should be fulfilled:

> someone who maintains a repo with a package that cares about cross compilation will have the option to switch completely as soon as they learn that they need to make a switch.

My concern with https://github.com/NixOS/nixpkgs/pull/237512 especially is that it introduces new API surface which becomes obsolete as soon as we remove functions from the platform set—requiring another breaking change to remove it.

**Comment by roberth** [comment]
> My concern with #237512 especially is that it introduces new API surface which becomes obsolete as soon as we remove functions from the platform set—requiring another breaking change to remove it.

Hardly anyone would have to use this function, because equality on `stdenv.*Platform` can work just fine for any single `stdenv`, by making sure that an identical value is put in both attributes, as early as possible when invoking Nixpkgs and/or constructing stdenv. See https://github.com/NixOS/nixpkgs/pull/237427#issuecomment-1588235885

Have we considered the alternative of disabling `==` equality altogether? Isn't `canExecute` a better relation to use anyway?

**Comment by amjoseph-nixpkgs** [comment]
> Construction is more accurate, it's not like it compares anything except pointers…

It should be possible to give semantics to a language which are independent of any particular implementation.  Obviously I've failed to do that, but I won't stop trying :smile:

> Have we considered the alternative of disabling `==` equality altogether?

I would love to do that but I don't think it's possible in Nix.

You can add an attribute whose value is `throw config` to prevent direct equality tests, but as soon as the user buries a platform inside some other attrset [Nix goes right back to pointer equality](https://code.tvl.fyi/tree/tvix/docs/value-pointer-equality.md#n116).  So you're still relying on Nix's weird function-equality -- it's just less obvious that you're relying on it.

**Comment by roberth** [comment]
> So you're still relying

Makes sense, so removing the functions is the right goal.
What remains is dealing with the breaking change better, so to make this PR step 1 of the migration, you could turn the placeholders back into working implementations.

**Comment by amjoseph-nixpkgs** [comment]
Adjusted the last commit message to make it conform to the nixpkgs standards.

**Review by roberth — DISMISSED** [review]
> The same restriction should apply to "mere" warnings as well.

We need to make the alternative - the new functions - widely available before adding any warnings.
As mentioned before, we shouldn't bother third party maintainers with this migration. Their migration should be instantaneous. That implies no warnings until it's time to perform the migration, because otherwise we'll both users, who will then bother the third party maintainers, who will implement the change, break the stable release, implement a version conditional, etc. This is just about the worst UX outcome, for an issue that's not even urgent.

Please first make a PR (this one or a new one) that does not remove the functions and does not change the functions' behavior.

**Comment by alyssais** [comment]
> Please first make a PR (this one or a new one) that does not remove the functions and does not change the functions' behavior.

I think a good compromise here would be to add the new functions, backport them to stable (which is valid because it's a purely additive change, like a package addition), and then add the deprecation warning on unstable a couple of days later.

That way, maintainers of out-of-tree code that should work on both stable and unstable don't have to maintain two separate code paths, resolving the issue @roberth is worried about, without this turning into a six month long project for @amjoseph-nixpkgs.

**Comment by amjoseph-nixpkgs** [comment]
> without this turning into a six month long project for @amjoseph-nixpkgs.

Yeah I have waaaay too many of those going on right now.

**Comment by roberth** [comment]
If you add the error, make very clear that users need to update stable to at least <commit> merged at <date>. Error messages should be actionable.
Nonetheless I would like to postpone the break to later this year.

> a six month long project

I would recommend to use a calendar. It's like a time travel.

**Comment by amjoseph-nixpkgs** [comment]
Rebased.

**Comment by amjoseph-nixpkgs** [comment]
Resolved merge conflict.

File: lib/systems/default.nix
Comment (alyssais) [inline lib/systems/default.nix:50]: If this has been deprecated for over a year, maybe we can just remove it now?

File: lib/systems/default.nix
Comment (alyssais) [inline lib/systems/default.nix:114]: Won't this be overridden by the definition of `emulator` below?  (and likewise for `emulatorAvailable`)

File: lib/systems/default.nix
Comment (alyssais) [inline lib/systems/default.nix:6]: Shouldn't we remove the original definition of this below?

Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:6]: 5dd051aa06a2

File: lib/systems/default.nix
Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:114]: Thanks for noticing.  This PR is very annoying to rebase.

5dd051aa06a2

File: lib/systems/default.nix
Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:50]: dd789192703c

File: lib/systems/default.nix
Comment (alyssais) [inline lib/systems/default.nix:112]: This could just be `emulator final`.  This attrset is non-recursive, so there's no need for the placeholder bindings.

**Review by alyssais — CHANGES_REQUESTED** [review]

File: lib/systems/default.nix
Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:112]: 9097e33a3d73

File: lib/systems/default.nix
Comment (alyssais) [inline lib/systems/default.nix:109]: These also don't need to be placeholders, right?

File: lib/systems/default.nix
Comment (alyssais) [inline lib/systems/default.nix:52]: And this is now unused.

File: lib/systems/default.nix
Comment (alyssais) [inline lib/systems/default.nix:87]: Why not just have the definitions here, instead of defining them above and then inheriting them into a recursive attrset?

File: lib/systems/default.nix
Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:109]: These were added by @sternenseemann here: https://github.com/NixOS/nixpkgs/pull/238331#pullrequestreview-1485078058 in commit 5d8e668d16f7ce7c32399007d3a49a27322ac3ed

And yes they do need to be placeholders, because they are functions.  And Nix's definition of equality for functions is crackheaded.  Which is why we are trying so desperately to remove the functions!

File: lib/systems/default.nix
Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:52]: See https://github.com/NixOS/nixpkgs/pull/238331#discussion_r1382315387

File: lib/systems/default.nix
Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:87]: 3c760ba7ab9a

File: lib/systems/default.nix
Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:107]: See https://github.com/NixOS/nixpkgs/commit/5d8e668d16f7ce7c32399007d3a49a27322ac3ed#diff-2165823a8d82c5dd1353601bd290df8bd431f9ee2096750d9ef655cf5d251998R37-R47

```nix
 # These definitions have to be floated out into a `let … in` binding to ensure
 # that they are the same exact value struct in the evaluating Nix
 # implementation. Due to a quirk in how equality is implemented, values including
 # functions are considered equal if they have the same memory location as long as
 # they are inside a container value and not compared directly. Thus any attribute
 # set will compare equal to itself even if it contains functions.
 # By floating the functions out of the attribute set construction we'll ensure
 # that the functions have the same memory location regardless of what call to
 # lib.systems.elaborate produced them. This ensures that platform sets will be
 # equal even if they have been constructed by different calls to lib.systems.elaborate
 # —as long as their other contents are the same.
```

**Comment by amjoseph-nixpkgs** [comment]
Rebased

File: lib/systems/default.nix
Comment (alyssais) [reply lib/systems/default.nix:138]: Ugh, right.

File: lib/systems/default.nix
Comment (alyssais) [inline lib/systems/default.nix:41]: The indentation for this whole bit is off.

**Review by alyssais — COMMENTED** [review]
That's my last comment — consider me to have approved the PR once that's fixed and the commits are squashed.

I've only reviewed the changes in lib/systems.  I'm trusting you to have done the mechanical replacement in the rest of the tree correctly.

File: lib/systems/default.nix
Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:41]: 653af6ae0869

**Comment by amjoseph-nixpkgs** [comment]
Rebased, squashed.

**Review by alyssais — APPROVED** [review]
Again, I've only taken a very cursory look outside of lib/systems.

File: lib/systems/default.nix
Comment (roberth) [inline lib/systems/default.nix:64]: Please change this to a warning.
By throwing, you force third party package expressions to insert compatibility logic in order to support the current releases.
This migration is not urgent, so we should not force third party packagers to write this kind of logic.

**Review by roberth — DISMISSED** [review]
~I wish I didn't have to [repeat](https://github.com/NixOS/nixpkgs/pull/238331#pullrequestreview-1487067630) myself.~ my bad, please ignore

File: lib/systems/default.nix
Comment (alyssais) [reply lib/systems/default.nix:64]: @roberth I think you are missing some context here.

As you can see in the diff, `isCompatible` is already a hard error, from May.  There's no point in this PR bringing it _back_.

File: lib/systems/default.nix
Comment (alyssais) [inline lib/systems/default.nix:41]: To be kind to external users supporting both stable and unstable, these shouldn't be made warnings until after 24.05 is branched off, and errors after 24.11 is branched off, right?

File: lib/systems/default.nix
Comment (roberth) [reply lib/systems/default.nix:64]: I tried to review too quickly, and I jumped to a wrong conclusion here based on the title and earlier discussion, but not enough reading.

File: lib/systems/default.nix
Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:41]: > To be kind to external users supporting both stable and unstable, these shouldn't be made warnings until after 24.05 is branched off, and errors after 24.11 is branched off, right?

I was hoping to get the `lib.warn` in before the 24.05 branch-off, since `lib.warn` is only a hard CI failure for `nixpkgs` itself.

Since `lib.warn` is our way of giving a heads-up to people downstream of us, anybody who has their CI set up to fail if `nixpkgs` emits a warning is basically turning anything sent across our only communication channel into breakage.  I think that's a bad choice and I don't think many users choose it (for very long).

So ideally I'd like to put the `lib.warn` in *before* 24.05.

The timeline for the removal of these attributes is less urgent; if that needs to wait until *after* 24.11 I guess that's okay.  Keep in mind however that until these attributes are removed, we have no way to `lib.warn` on any deprecated attributes of system objects (like `isEfi`).

Ultimately, if people really want even the `lib.warn` to wait until after 24.05, then I guess it has to wait.

Even though this PR has two approvals I will not merge it myself.

Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix]: Alright, let me know if 1f7f6d396241c54829df4209c383a1aa2062c35a captures this intent correctly.

**Comment by amjoseph-nixpkgs** [comment]
Fixed merge conflict.

File: lib/systems/default.nix
Comment (roberth) [inline lib/systems/default.nix:66]:
```suggestion
  # uncomment in January 2024, after 23.05 EOL
```

File: lib/systems/default.nix
Comment (roberth) [inline lib/systems/default.nix:67]:
```suggestion
  # delete canExecutePlaceholder, emulatorAvailablePlaceholder and emulatorPlaceholder after 24.05 branch-off (throwing would have been nice, but throws off `==` equality when comparing platforms elaborated on distinct but compatible Nixpkgs source trees)
```

- Won't be grouped by the comment syntax anymore, so be explicit
- Explain why not throw in this case

File: lib/tests/systems.nix
Comment (roberth) [inline lib/tests/systems.nix:68]: I think this needs to be in sync with the other uncomment comment.

```suggestion
# Uncomment after 23.05 EOL in January 2024
```

**Review by roberth — COMMENTED** [review]
We should move the first date a bit, because for a month we support all three versions: 23.05, 23.11, unstable.

Maybe it'd be more clear to make the 23.05 EOL changes in a draft PR instead of commented code.
Perhaps link the PR in a comment if you still want to provide context for readers of the code.

**Pull request closed without merging**
